content/src/tnips/tnip_2.md

```diff
@@ -0,0 +1,279 @@
+
+Introducing a process to rotate eligible validators provides fresh opportunities for new nodes to participate in consensus as voting committee members. Randomly shuffling validators underpins both network security and community growth, aligning with the broader goals of decentralized security and equitable participation.
+
+## Specification
+##### Handshake
+
+The handshake is permissionless. [NVVs](#non-voting-validators-nvvs) initiate contact with known validators to subscribe to the latest consensus. The handshake involves exchanging network information to establish peer connections.
+
+### Committees
+
+Consensus is managed by nodes that staked TEL. These nodes form committees and cast votes to reach a quorum. Committees are only valid for the epoch in which they are current.
+
+##### Committee Voting Validators ([CVVs](#committee-voting-validators-cvvs))
+
+Validators currently in the committee, responsible for casting votes, extending the canonical chain, and reaching consensus on transactions and finality.
+
+##### Non-Voting Validators ([NVVs](#non-voting-validators-nvvs))
+
+Validators that track and execute consensus but do not vote for every block in the epoch. They participate by receiving consensus through a gossip network consisting of both [CVVs](#committee-voting-validators-cvvs) and [NVVs](#non-voting-validators-nvvs). [NVVs](#non-voting-validators-nvvs) only vote on the latest execution result to support new committees during epoch transitions.
+
+##### On-chain Committee Information
+
+Committee information is stored on-chain to support client verification. The “Consensus Registry” contract is located at 0x467Bccd9d29f223BcE8043b84E8C8B282827790F and includes the following committee types:
+
+* C: The current committee which votes to reach consensus and extend the canonical tip.
+* C<sub>p</sub>: The committee from the previous epoch.
+* C<sub>n</sub>: The committee in the next epoch.
+* C<sub>n+1</sub>: The committee after the next epoch, which is 2 epochs ahead of the current epoch.
+
+##### Committee Transition
+
+At each epoch boundary, the outgoing committee signs and broadcasts the last sealed execution result. The incoming committee waits for a quorum of signed execution results to establish the genesis certificate for the new epoch.
+
+### Epochs
+
+An epoch is defined as a 24-hour time period. The transition is triggered by the first commit to the consensus Directed Acyclic Graph (DAG) after the 24-hour interval.
```

Comment:
I think we have to tether epochs to timestamps since the number of blocks per epoch is not deterministic and rounds share this relationship so neither can be used to define epoch boundaries, so I think we will have to devote a good amount of thought to timing attacks. There is a good amount of ethereum research on this but a lot of it won't be relevant to our schema

Another thing is that we will need to make sure onchain developers are made aware of this in our documentation because Solidity and Vyper best practices recommend using block numbers to measure the progression of time (even under PoW this was recommended). This is because unix timestamp is manipulable to some extent. 

Eg if for whatever reason the engine were ordering transactions by timestamp for example, validators could insert sandwich attack transactions or if there were a betting dapp programmed to use `block.timestamp` a validator could report a malicious block timestamp and mine one just-in-time with their own bet at the top of the block

Reply:
I'm glad you're thinking about these types of security concerns!

I think you bring up a really good point that we need to encourage devs to prefer block height over timestamps as a measure of time.

Here's my assessment of why I think users are better protected from these types of timing attacks:

Transactions are sequenced at the worker level when they enter the mempool. During the construction of a worker's batch, it is possible to sandwich attack however it would be obvious which node was responsible. At that point, users could avoid the node all together and governance could act.

Once a worker block reaches quorum, it is included in the Primary's header for a round. The primary immutably orders its batches by time. Once the primary's header reaches quorum, it is certified and eventually committed to the DAG. The DAG sequences primary block which ultimately sequences transactions.

Worker blocks are then executed in order, after consensus, using the committed subdag's timestamp. This means multiple blocks from the same wave of consensus output will have the same timestamp.